### ISSUE TITLE
Add hook to transform plugin list before registration

### ISSUE THREAD

**Issue by devinivy** [hapijs/glue#114]
This dates back to #35, but I propose we approach it differently this time around, in a way that should work better for everyone.

My request is a hook to transform the `plugins` argument that is passed to `server.register()`.  I think it could have several applications.  One thing that is so great about glue is that it creates a semi-formal point in a hapi app's lifecycle—one should be able to assume that the plugins in a glue manifest constitute the _entire_ list of top-level plugins registered in the application, and once `Glue.compose()` is complete, there are no more plugins to register.  This provides a unique opportunity to reorder plugin registrations that depend on each other.  I believe there are some valid cases where managing plugin registration order is necessary, e.g. for plugins that must be configured once with plugin options to provide shared functionality to several other plugins.

I propose adding an option to `compose()` called something like `pluginsTransform` with the signature `(plugins, server) => plugins`.  I would be happy to provide a PR if this sounds interesting.

**Comment by WesTyler** [comment]
I like this. I've been thinking around ways that it could screw things up for the past couple of days and haven't come up with anything significant. :)

I'll take a stab at this next week if you (or someone else) hasn't made a PR at that point.

**Comment by csrl** [comment]
I don't follow.  You have control over the manifest passed to glue, process it before calling compose().  Why do you need a function to transform what you yourself passed to glue in the first place?

Can you provide an actual example of why this is needed?

**Comment by devinivy** [comment]
The difference is primarily that glue does [some work](https://github.com/hapijs/glue/blob/master/lib/index.js#L86-L100) to resolve plugins.  I'm hoping to be able to integrate [`Hodgepodge.sort()`](https://github.com/devinivy/hodgepodge/blob/master/API.md#hodgepodgesortplugins-loosetally) and glue together seamlessly, but I think there could be other applications.

**Comment by csrl** [comment]
If I follow correctly, you are making configuration driven dependency management, satisfied through plugin registration ordering.

But you want plugin authors to write their plugins with support for this, which only works then if using glue (assuming the functionality is merged) or hodgepodge.  But would break/not work for any other method of loading the plugin.  Whereas if they use the built in hapi dependency management, it works everywhere.

Am I missing something?

**Comment by csrl** [comment]
If I ignore the reasoning behind the request, I can see a nicety in updating the glue compose flow to provide a hook between the plugin parsing and calling register.  And if such were done, I think it would make sense to generalize the hook functions to pass and return the manifest at each stage.  That would give the opportunity to manipulate the manifest in any fashion desired.

`onPreServer(server, manifest)` - must return manifest
`onPreRegister(server, manifeset)` - must return manifest

would be the two hooks, with `onPreServer` going between `internals.parseServer` and `Hapi.server()` and `onPreRegister` going after the `internals.parsePlugin` map and before `server.register`

It'd be a breaking change, as onPreRegister would happen after the plugins are parsed, whereas it currently happens before they are parsed.

If that satisfies your needs, I'm open to such a PR.

**Comment by csrl** [comment]
On second thought, the positioning of onPreRegister shouldn't affect existing code, and if the hook doesn't return a manifest we can continue with the existing one.

**Comment by devinivy** [comment]
Yeah, I'm not interested in adding hodgepodge's functionality to glue.  Instead I'm interested in a hook so that someone could write something like,
```js
const server = await Glue.compose(plugins, {
    pluginsTransform: Hodgepodge.sort // (plugins) => reorderedPlugins
});
```

Can you help me understand how you picture `onPreRegister` affecting glue's resolved `plugins`?

<details/><summary>To keep the thread tidy, I've written the impetus for hodgepodge in here :) Always happy to move to hapijs/discuss to extend the conversation!</summary>

There are _rare_ cases that hapi's dependency management doesn't quite do the trick, and then you're on your own.  This happens specifically when you depend on a plugin at registration time that also takes registration options.  Here's [an article](https://github.com/devinivy/hapipal.com/blob/master/lib/content/best-practices/handling-plugin-dependencies.md) I wrote on it, if you're interested in my line of thinking.

I find it happens often that when people start writing their own "internal" hapi plugins for a big project, at some point they start implicitly relying on plugin registration order.  Hodgepodge is meant to help mitigate that kind of technical debt as well.
</details>

**Comment by csrl** [comment]
```javascript
const server away Glue.compose(manifest, {
    onPreRegister: (server, manifest) => {
        manifest.register.plugins = Hodgepodge.sort(manifest.register.plugins);
        return manifest;
    });
});
```
or if you depend on side affects, assuming Hodgepodge modifies the plugins array in place.
```javascript
const server away Glue.compose(manifest, {
    onPreRegister: (server, {register: {plugins}}) => {
        Hodgepodge.sort(plugins);
    });
});
```

**Comment by csrl** [comment]
As for the linked article about dependencies.  I don't understand the registration time vs run time distinction.  There is no point in using server.dependency without placing the rest of the init code in side of the 'after' callback that requires the dependency.  It like hitting your finger with a hammer and saying the hammer is broken.  Use the tool properly.  But you obviously understand that, as the hodgepodge API.md example "Without hodgepodge" demonstrates correct usage of the feature.

Anyway, I'm open to a PR with my proposed approach.

**Comment by csrl** [comment]
@devinivy can you confirm that my proposal provides what you need?  I might have time to work on this in the near future.

**Comment by hueniverse** [comment]
6 months later, is there a reason this is still open?

**Comment by csrl** [comment]
It doesn't bother me that its sitting open or if it gets closed.  I'm not going to get to implementing the functionality any time soon.   If someone wants to create a PR for it, that'd be great, and with it sitting open maybe it catches someones attention at some point.  If you want it closed, that can happen.

**Comment by Nargonath** [comment]
If nobody's available to work on it, I guess it'd a good candidate for `backlog` label.

**Comment by devinivy** [comment]
I may be able to pick this up at some point, but it's relatively low priority for me, so not in the immediate future.  I'll close this—perhaps we can reopen it if a similar request comes through, or if it becomes a priority for some reason.  It's good/useful to know from this conversation that a PR would be welcome.
